stackrox/stackrox
# ROX-25725: Fix accessibility of Popover footer, header, and heading

### Description

Apply 80/20 principle to fix generic issues so team members can divide and conquer pages that have specific issues.

Work around PatternFly issue: https://github.com/patternfly/patternfly-react/issues/10852

### Problems reported by axe DevTools

> Heading levels should only increase by one

https://dequeuniversity.com/rules/axe/4.9/heading-order

> Document should not have more than one banner landmark

https://dequeuniversity.com/rules/axe/4.9/landmark-no-duplicate-banner

> Document should not have more than one contentinfo landmark

https://dequeuniversity.com/rules/axe/4.9/landmark-no-duplicate-contentinfo

> Ensures landmarks are unique

https://dequeuniversity.com/rules/axe/4.9/landmark-unique

### Analysis

1. Popup `headerComponent` has default `h6` element.
    https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Popover/PopoverHeader.tsx#L28
    But it is questionable whether popup should be in heading hierarchy of page.
2. Popup `headerContent` renders `header` element in addition to `MastHead` element.
    https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Popover/PopoverHeader.tsx#L34
3. Popup `footerContent` renders `footer` element in addition to `MainPage` element.
    https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Popover/PopoverFooter.tsx#L17
    https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/MainPage/MainPage.tsx#L110-L112
4. Similar to items 2 and 3.

### Solution

1. Edit pluginAccessibility.js file.
    * Add positive `Popover-aria-label-prop` rule.
    * Add negative `no-Popover-footerContent-headerContent-props` rule.
2. Add PopoverBodyContent.tsx file.
    * Compose body with header, or footer, or both.
    * Replace `header` and `footer` with `div` elements.
    * Replace heading with `p` element.
    * Use PatternFly classes for compatible appearance.
3. Edit acs.css file.
    * Add style rules when header as child of body.
4. Edit trumps.css file.
    * Add selectors for link in `Popover` element (left-over from PatternFly 5 upgrade).
5. Edit file that have errors for 2 new lint rules.
    * Add `aria-label` prop if needed.
    * Replace props with helper element.

### User-facing documentation


### Testing and quality


#### Automated testing


#### How I validated my change

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js 0 = 4618273 - 4618273
        total 4648 = 11709592 - 11704944
    * `ls -al build/static/js/*.js | wc`
        files 0 = 175 - 175
3. `yarn start` in ui

#### Manual testing

1. Visit /main/dashboard, and then click gear at the right of **Images at most risk**.

    * Before changes, see presence of accessibility issue for **name**.
        ![WidgetOptionsMenu_without_aria-label](https://github.com/user-attachments/assets/273b7a2a-4e30-45e9-9a68-dbd89e9f3d0f)

    * After changes, see absence of accessibility issue for **name**.
        ![WidgetOptionsMenu_with_aria-label](https://github.com/user-attachments/assets/43efde31-923e-42fa-82df-5e506aa50413)

2. Visit /main/network-graph, select cluster and namespaces, click **Network policy generator**, click **Generate and simulate network policies**, and then click question mark at the right of **Compare**.

    * Before changes, see presence of accessibility issue.
        ![NetworkPolicySimulatorSidePanel_with_issues](https://github.com/user-attachments/assets/4fea9141-3ff1-4c5c-bffd-928426a8778a)

    * After changes, see absence of accessibility issue.
        ![NetworkPolicySimulatorSidePanel_without_issues](https://github.com/user-attachments/assets/b77b1db9-d358-4ceb-8186-711e42415783)

3. Visit /main/vulnerabilities/workload-cves, select **View images without vulnerabilities**, and then click **Image scanning incomplete** label.

    * Before changes, see presence of 3 accessibility issues for **heading** and **landmark**.
        ![ImageScanningIncompleteLabel_with_issues](https://github.com/user-attachments/assets/d15b18e8-1a99-417a-bda1-3e058d955177)

    * After changes, see absence of 3 accessibility issues for **heading** and **landmark**.
        ![ImageScanningIncompleteLabel_without_issues](https://github.com/user-attachments/assets/b0cd9061-afdf-4260-b074-1799d95f1030)

        ![CVE_data_may_be_inaccurate_Elements](https://github.com/user-attachments/assets/3d211145-9b66-4a6b-a798-51b0e5a2c799)

4. Visit /main/integrations/authProviders/machineAccess/create, click **Add new rule**, and then click question mark at the right of **Value**.

    * Before changes, see presence of accessibility issues.
        ![MachineAccessIntegrationForm_with_issues](https://github.com/user-attachments/assets/1e3a5171-2186-4639-98ea-c3f35d492939)

    * After changes, see absence of accessibility issues.
        ![MachineAccessIntegrationForm_without_issues](https://github.com/user-attachments/assets/11a278cf-e629-4562-8e52-c7b522fc13e8)

        ![MachineAccessIntegrationForm_Elements pmg](https://github.com/user-attachments/assets/6ff61f55-a8c2-42f8-85de-82cb23140baa)

5. Visit /main/system-health, and then click **Generate diagnostic bundle**.

    * Before changes, see presence of accessibility issues.
        ![GenerateDiagnosticBundle_with_issues](https://github.com/user-attachments/assets/0bbe826b-f9d1-4f25-a291-3bcda3d4b825)

    * After changes, see absence of accessibility issues.
        ![GenerateDiagnosticBundle_without_issues](https://github.com/user-attachments/assets/4cf81023-50b1-4341-91c5-887e0024c33f)

        ![GenerateDiagnosticBundle_Elements](https://github.com/user-attachments/assets/7a33ccc1-f8bc-49ce-b450-26bdc1f91a92)

#### Integration testing

* clusters/clustersHealthStatus.test.js
* integrations/externalBackups.test.js
* integrations/imageIntegrations.test.js
* integrations/machineAccessConfigs.test.js
* integrations/notifiers.test.js
* integrations/signatureIntegrations.test.js
* systemHealth/bundle.test.js
* vulnerabilities/exceptionManagement/approvedDeferralsTable.test.js
* vulnerabilities/exceptionManagement/approvedFalsePositiviesTable.test.js
* vulnerabilities/exceptionManagement/approveRequestFlow.test.js
* vulnerabilities/exceptionManagement/cancelRequestFlow.test.js
* vulnerabilities/exceptionManagement/deniedRequestsTable.test.js
* vulnerabilities/exceptionManagement/denyRequestFlow.test.js
* vulnerabilities/exceptionManagement/general.test.js
* vulnerabilities/exceptionManagement/pendingRequestsTable.test.js
* vulnerabilities/exceptionManagement/requestDetails.test.js
* vulnerabilities/exceptionManagement/updateRequestFlow.test.js
